### PR TITLE
REGR: Fix interpolation on empty dataframe

### LINKED ISSUE #35598: BUG: interpolate gives a TypeError on empty dataframes
Already fixed in #35543, making an issue for linking/discussion.

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
from pandas import DataFrame
DataFrame().interpolate()
```
    "Cannot interpolate with all object-dtype columns "
TypeError: Cannot interpolate with all object-dtype columns in the DataFrame. Try setting at least one column to a numeric dtype.

#### Problem description
This is a regression, 1.0.5 handles it fine.

The problem appeared due to checking `np.all(obj.dtypes == np.dtype(object))` since the all of an empty set is True

#### Output of ``pd.show_versions()``

pandas           : 1.1.0

### PR DESCRIPTION
**Pull request by sanderland**
Interpolation on an empty dataframe broke in 1.1 due to a change in how 'all columns are objects' is checked (specifically all(empty set) is True, while before dtype count object = None was checked against size = 0).
This is a complex function and I'm not sure what the proper fix is, suggesting to keep the empty check out of the rest of the logic.

Example code that broke:

```python
import pandas as pd
df = pd.DataFrame([1,2])
df[[]].interpolate(limit_area='inside')
```

- [x] closes #35598
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
Can you add a test case for this?

**Comment by sanderland** [comment]
> Can you add a test case for this?

done

File: pandas/tests/frame/methods/test_interpolate.py
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:38]: can you add the issue number as a comment here

File: pandas/tests/frame/methods/test_interpolate.py
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:39]: can you do
```
result = 
expected = 
tm.assert_frame_equal(result, expected)
```

File: pandas/core/generic.py
Comment (jreback) [inline pandas/core/generic.py:6802]: would prefer that this happens in the internal method itself (called on L6861)

**Review by jreback — CHANGES_REQUESTED** [review]
i guess this was a regression, can you add a note in 1.1.1

File: pandas/core/generic.py
Comment (sanderland) [reply pandas/core/generic.py:6802]: The error happens before this call (on line 6825)

File: doc/source/whatsnew/v1.1.1.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.1.1.rst:20]:
```suggestion
- Fixed regression where :meth:`DataFrame.interpolate` would raise a ``TypeError`` when the :class:`DataFrame` was empty (:issue:`35598`).
```

File: pandas/tests/frame/methods/test_interpolate.py
Comment (simonjayhawkins) [reply pandas/tests/frame/methods/test_interpolate.py:38]: can you do this as a code comment, e.g.
```suggestion
        # https://github.com/pandas-dev/pandas/issues/35598
        df = DataFrame()
```

File: pandas/core/generic.py
Comment (simonjayhawkins) [reply pandas/core/generic.py:6810]: There were a few changes to interpolate in 1.1.0. will run git bisect to ascertain why this is now failing.

Comment (simonjayhawkins) [reply pandas/core/generic.py:6810]: There was a subtle change in #34752 where

```
if self.ndim == 2 and np.all(self.dtypes == np.dtype(object)):
```
was changed to 
```
if obj.ndim == 2 and np.all(obj.dtypes == np.dtype(object)):
```

does reverting this change restore the old behaviour.

Comment (simonjayhawkins) [reply pandas/core/generic.py:6810]: the regression is from #33084, but since the error message is about DataFrame columns, I don't think the above should have been changed either. @jbrockmendel

Comment (jorisvandenbossche) [reply pandas/core/generic.py:6823]: > would prefer that this happens in the internal method itself (called on L6861)

As @sanderland notes, the check for all object dtype indeed happens here before calling into the internal method, so the check for empty thus also needs to happen here

However, @sanderland I think the `df.empty` check is not fully correct. Because this was also give True if you have columns but no rows. And for example an empty dataframe with a datetime64[ns] column will give a timedelta64[ns] column as result. That's something we should keep.

So I think we should explicitly check for no columns / no rows (alternatively could also add the check to the offending `if obj.ndim == 2 and np.all(obj.dtypes == np.dtype(object)):`, eg `.. and obj.shape[0]`)

Comment (jreback) [reply pandas/core/generic.py:6823]: @sanderland can you try @jorisvandenbossche suggestion here

Comment (sanderland) [reply pandas/core/generic.py:6823]: > However, @sanderland I think the `df.empty` check is not fully correct. Because this was also give True if you have columns but no rows. And for example an empty dataframe with a datetime64[ns] column will give a timedelta64[ns] column as result. That's something we should keep.

I tried this in 1.0.x and it does not, I dont see why interpolate would do this either. Could you suggest a test which fails with my approach but passes in 1.0.x?

Comment (jorisvandenbossche) [reply pandas/core/generic.py:6823]: Hmm, you're fully correct. I might have been mixing up the review of two PRs, as I was also reviewing a regression for `diff()` and maybe was testing that method here as well .. ;) (since `df.diff()` for datetime64 gives timedelta64, also for empty dataframe)

Forget my comment!

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@sanderland can you merge master once more to see if that fixes CI?

**Comment by simonjayhawkins** [comment]
test failure was

```
=========================== short test summary info ===========================
FAILED pandas/tests/computation/test_eval.py::TestAlignment::test_basic_series_frame_alignment[numexpr-python]
= 1 failed, 72824 passed, 1746 skipped, 1055 xfailed, 58 warnings in 1102.57s (0:18:22) =
```
probably unrelated.

restarted (azure will patch against master)

**Comment by simonjayhawkins** [comment]
Thanks @sanderland

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Review by jreback — COMMENTED** [review]
@simonjayhawkins in the future pls don't merge these so fast

this is not what we wnat here

eg this should actually return self.copy()

not to mention this is not the correct place for this

**Comment by jorisvandenbossche** [comment]
I think Simon's call to merge this was appropriate (all the existing review comments were addressed, it was reviewed by several core devs). Yes, you noticed an additional problem (the missing copy) after merge: no problem, that can always happen, and we can do a follow-up to fix this.

> not to mention this is not the correct place for this

See my comment at https://github.com/pandas-dev/pandas/pull/35543#discussion_r469979874 (and @sanderland's own answer above that), your suggestion to move it down into the internals is not possible, so this function is a correct place to put the check.

**Pull request merged**
